[Guided onboarding] Landing page updates

yuliacech opened the pull request:
## Summary
Fixes https://github.com/elastic/kibana/issues/138548 

This PR updates the Guided onboarding Landing page that can be found at `/app/home#/getting_started`. The Landing page has now 4 cards: 3 solution guides and a link to Infrastructure category. The guide card displays the state of each guide with the progress bar and a button to "start/continue" the guide.

Note: The "view integrations" button links to `/app/integrations/browse/infrastructure` but the category doesn't have any integrations added to it yet. As soon as there are any integrations in that category, the link should work as expected.

### Screenshots 
#### No guides started yet
The header button is disabled, guide cards have "view guide" buttons.

<img width="1786" alt="Screenshot 2022-10-19 at 18 46 32" src="https://user-images.githubusercontent.com/6585477/196754743-0fb32db3-db97-4d99-8e5a-ad35db2472b2.png">


#### "View guide" button is clicked
The dropdown panel shows the selected guide and the user can start the first step. The header button is enabled. 

<img width="1781" alt="Screenshot 2022-10-19 at 18 46 56" src="https://user-images.githubusercontent.com/6585477/196754776-b5d7ead0-7f03-43bf-8608-f9161498da19.png">


#### A guide is in progress
The guide card displays the progress and the "continue" button.

<img width="302" alt="Screenshot 2022-10-19 at 18 47 27" src="https://user-images.githubusercontent.com/6585477/196754835-322e295c-85cf-47c7-8a0d-e1e9795136b2.png">


#### A guide has been completed
The guide card displays the completed progress bar and the "view guide" button.

<img width="308" alt="Screenshot 2022-10-19 at 18 48 27" src="https://user-images.githubusercontent.com/6585477/196754863-400ec3b3-9c44-4a43-8b53-b29ca1a86aef.png">


### How to test
1. Enable the guided onboarding plugin by adding `guidedOnboarding.ui: true` to the file `/config/kibana.dev.yml`.
2. Start ES and Kibana with examples: `yarn es snapshot` and `yarn start --run-examples`.
3. Navigate to `/app/home#/getting_started` and check the landing page when no guides have been started yet.
4. Click the card for the search guide and check that the dropdown panel is being displayed and the first step can be started.
5. Start the first step and complete it on the page `/app/guidedOnboardingExample/stepOne`.
6. Navigate back to the landing page and check that the search card now displays the progress. 
7. Complete the search guide and check that the landing page displays the search guide as completed.   

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

elasticmachine commented:
Pinging @elastic/platform-onboarding (Team:Journey/Onboarding)

cindychangy reviewed (approved):
Thanks @yuliacech, this looks good! I left a couple of small style tweaks. They can move to the larger design polish task as well if preferred.

cindychangy commented on src/plugins/home/public/application/components/guided_onboarding/use_case_card.tsx:
```diff
@@ -113,37 +70,29 @@ export const UseCaseCard = ({ useCase }: UseCaseProps) => {
-
-  const title = (
+  const titleElement = (
     <EuiTitle size="xs">
```
```suggestion
    <EuiText textAlign="center">
```
We want to center the contents of the card, oddly when applying `center` to the `align` prop on the parent `EuiCard` it centers the progress label as well, I think this is the cleanest way to achieve the centered text while keep the progress left aligned?

cindychangy commented on src/plugins/home/public/application/components/guided_onboarding/use_case_card.tsx:
```diff
@@ -113,37 +70,29 @@ export const UseCaseCard = ({ useCase }: UseCaseProps) => {
-        <strong>{constants[useCase].i18nTexts.title}</strong>
+        <strong>{title}</strong>
       </h4>
     </EuiTitle>
```
```suggestion
    </EuiText>
```

cindychangy commented on src/plugins/home/public/application/components/guided_onboarding/use_case_card.tsx:
```diff
@@ -113,37 +70,29 @@ export const UseCaseCard = ({ useCase }: UseCaseProps) => {
   );
-  const description = (
+  const descriptionElement = (
     <EuiText color="subdued" size="xs">
```
```suggestion
    <EuiText size="s" textAlign="center">
```

cindychangy commented on src/plugins/home/public/application/components/guided_onboarding/getting_started.tsx:
```diff
@@ -86,15 +99,18 @@ export const GettingStarted = () => {
           <EuiSpacer size="s" />
           <EuiSpacer size="xxl" />
-          <EuiFlexGrid columns={3} gutterSize="xl">
+          <EuiFlexGrid columns={4} gutterSize="l">
```
I was unable to insert directly on line `97` but can we slightly modify the line to read: `<EuiText size="m" textAlign="center">`

yuliacech commented:
Thanks a lot for your review, @cindychangy!

majagrubic commented on src/plugins/home/public/application/components/guided_onboarding/guide_card.tsx:
```diff
@@ -0,0 +1,196 @@
+    </>
+  );
+};
+export const GuideCard = ({ useCase, guides }: { useCase: UseCase; guides: GuideState[] }) => {
```
Could this component be a separate component in a `shared-ux` package?

majagrubic commented on src/plugins/home/public/application/components/guided_onboarding/guide_card.tsx:
```diff
@@ -0,0 +1,196 @@
+  },
+};
+
+const getCardFooter = (
```
It feels the footer requires a component on its own.

majagrubic commented on src/plugins/home/public/application/components/guided_onboarding/guide_card.tsx:
```diff
@@ -0,0 +1,196 @@
+  if (!guideState || guideState.status === 'not_started') {
+    return viewGuideButton;
+  }
+  const numberSteps = guideState.steps.length;
```
nit: can we destructure `guideState` to get `steps` and `state`?

majagrubic commented on src/plugins/home/public/application/components/guided_onboarding/guide_card.tsx:
```diff
@@ -0,0 +1,196 @@
+          )}
+        />
+        <EuiSpacer size="l" />
+        <EuiFlexGroup justifyContent="center">
```
do you really need a flexGroup with just one item?

majagrubic reviewed (changes requested):
Thank you for doing this. I didn't have a thorough look, but I feel these components should be a separate package, rather than live inside a home plugin. I am not sure if they should live under `shared-ux` space (depends how or when are they going to be reused) or if they should be under `onboarding`.

alisonelizabeth reviewed (approved):
This looks great @yuliacech 🎉 

I left a few comments in the code, nothing blocking. I'm going to approve, but if you don't mind taking a look before merging and let me know what you think.

I also found a bug with the "Continue" step button. It doesn't appear to be related to your changes, so I went ahead and opened https://github.com/elastic/kibana/issues/143306 to address separately.

majagrubic commented:
@yuliacech hopefully you can use this presentation as a guideline:
https://docs.google.com/presentation/d/1Za98KuAI0S7DQrVR-FGMj-H62er7j8YmNY2rCe7PYHI/edit

and then you can let me know if you have further questions.
You can find some good examples under `shared-ux` in packages.

TinaHeiligers commented on packages/kbn-guided-onboarding/src/components/landing_page/guide_card.test.tsx:
```diff
@@ -0,0 +1,42 @@
+
+import { GuideCard } from './guide_card';
+
+jest.mock('../../kibana_services', () => {
```
You'll need to refactor the test mock to import and mock out dependencies.

yuliacech replied:
Yes, I haven't completed the refactoring yet. I'm concerned that in the plugin I can't import the code out of the package. There is an error that the module doesn't exist.

yuliacech replied:
Oh never mind, I think I figured it out :)

TinaHeiligers commented on packages/kbn-guided-onboarding/src/components/landing_page/guide_card.test.tsx:
```diff
@@ -0,0 +1,42 @@
+import { GuideCard } from './guide_card';
+
+jest.mock('../../kibana_services', () => {
+  const { uiSettingsServiceMock, httpServiceMock } = jest.requireActual('@kbn/core/public/mocks');
```
These can be imported directly from their packages:
`uiSettingsServiceMock` from `@kbn/core-ui-settings-browser-mocks`
`httpServiceMock` from `@kbn/core-http-browser-mocks`

yuliacech commented on src/plugins/guided_onboarding/public/services/api.ts:
```diff
@@ -121,6 +121,12 @@ export class ApiService implements GuidedOnboardingApi {
+    // deactivate the current active guide if any
+    const activeGuide = await firstValueFrom(this.fetchActiveGuideState$());
+    if (activeGuide && activeGuide.guideId !== guideId) {
+      await this.deactivateGuide(activeGuide);
```
I didn't know that happens automagically :)

yuliacech commented on src/plugins/home/public/application/components/guided_onboarding/guide_card.tsx:
```diff
@@ -0,0 +1,196 @@
+  useCase: UseCase,
+  guidedOnboardingService?: GuidedOnboardingApi
+): ReactNode => {
+  const guideState = guides.find((guide) => guide.guideId === (useCase as GuideId));
```
I think we still would have to iterate over the state for each card because we have an array in the `getting_started` component.

yuliacech commented:
Thanks a lot for your review, @alisonelizabeth! Appreciate your comments in the code 👍

yuliacech commented:
Thanks a lot for your review, @majagrubic! I moved the code out of the home plugin into the package, would you mind having another look please?

yuliacech commented:
Thank you so much for having a look, @TinaHeiligers! I think I was able to figure out the package setup, fingers crossed the CI will be green.

majagrubic commented on src/plugins/home/public/application/components/guided_onboarding/getting_started.tsx:
```diff
@@ -81,20 +100,45 @@ export const GettingStarted = () => {
             </EuiFlexItem>
             <EuiFlexItem>
-              <UseCaseCard useCase="security" />
+              <GuideCard
```
Can we render these cards in a loop? It looks like (almost) all are the same

yuliacech replied:
We need to display a guide card for search, a guide card for observability, a link card for observability and a guide card for security, so I don't think a loop would work with this logic.

majagrubic replied:
Would something like this not work?
```
{ ['search', 'observability', 'observabilityLink', 'security'].map((useCase) => {
      if (useCase === 'observabilityLink') { 
         return (<EuiFlexItem>
                <ObservabilityLinkCard
                  ....
                />
            </EuiFlexItem>) ;
      } 
      return (<EuiFlexItem>
         <GuideCard useCase={useCase} .../>
       </EuiFlexItem>);
    };
}
```

majagrubic reviewed (commented):
Thanx for extracting this into a package. I left some comments around the code and the home plugin specifcially.

yuliacech commented on src/plugins/home/public/application/components/guided_onboarding/getting_started.tsx:
```diff
@@ -73,6 +86,12 @@ export const GettingStarted = () => {
+  const isDarkTheme = uiSettings.get<boolean>('theme:darkMode');
+  const activateGuide = async (useCase: UseCase, guideState?: GuideState) => {
+    await guidedOnboardingService?.activateGuide(useCase as GuideId, guideState);
+    // TODO error handling
```
We don't have the correct error handling in the API yet, so this will be addressed in a separate [issue](https://github.com/elastic/kibana/issues/139798)

majagrubic replied:
Could you link the issue in the TODO then?

yuliacech commented on packages/kbn-guided-onboarding/src/components/landing_page/guide_card_footer.tsx:
```diff
@@ -0,0 +1,120 @@
+          )}
+        />
+        <EuiSpacer size="l" />
+        <div className="eui-textCenter">
```
I followed your suggestion to not use a flexGroup when there is only 1 component. Please let me know if there is a better way to center the button in the footer.

majagrubic replied:
Sorry,  maybe my comment was clear initially. We shouldn't be using `eui-` classes directly in our React components, as they are internal to EUI and could possibly change. I meant - you can use `EuiFlexGroup` to achieve this (it's fine even for one component). Another way is to set the `div` width to 50% and  margin to `auto`:
https://jsfiddle.net/79kjtwqo/1/

alisonelizabeth replied:
I don't think `eui-textCenter` is internal to EUI, or at least is is documented as an available utility class here - https://elastic.github.io/eui/#/theming/typography/utilities.

yuliacech replied:
I agree with Alison here, if the EUI class should not be used in Kibana, there should be an eslint rule for that.

clintandrewhall replied:
@majagrubic was absolutely correct here.

@alisonelizabeth it's not that it's internal to EUI, it's that they're likely not going to exist for much longer.  Even on the [documentation link](https://elastic.github.io/eui/#/theming/typography/utilities) you reference, it states:

>For most of your usages we recommend using [EuiText](https://elastic.github.io/eui/#/display/text) or [EuiTitle](https://elastic.github.io/eui/#/display/title) components directly instead of these theme tokens. Or head to the [text utilities page](https://elastic.github.io/eui/#/theming/typography/utilities) for helper classes and functions.

EUI is moving to Emotion:

> EUI is transitioning to a CSS-in-JS solution for theming and requires consuming/customizing global variables in both the Sass and CSS-in-JS methods. To track this effort, subscribe to the [Meta ticket](https://github.com/elastic/eui/issues/3912).

Using these classes directly should be for edge cases only.  The right solution for this would have been to use `EuiText`, rather than a `div` with a global EUI classname.

@yuliacech The reason there's no lint rule is that it would literally blow up the linter, due to the number of exceptions we've made in the past, (or older cases before Emotion and theming).

I would recommend a follow-up PR to switch to `EuiText`.

alisonelizabeth replied:
> @alisonelizabeth it's not that it's internal to EUI, it's that they're likely not going to exist for much longer. 

👍 thanks for clarifying

> The right solution for this would have been to use EuiText, rather than a div with a global EUI classname.

This is for centering a button though, right @yuliacech? I don't _think_ `EuiText` is appropriate in that case. Perhaps @majagrubic earlier suggestion is the best approach:

> What I meant was - you can use EuiFlexGroup to achieve centering of the button (it's fine even for one component). Another way is to set the div width to 50% and margin to auto:
> https://jsfiddle.net/79kjtwqo/1/

I've opened https://github.com/elastic/kibana/issues/143794 to follow up on this.

yuliacech replied:
Thanks a lot for the context, @clintandrewhall! I just wish this were communicated more broadly because with the number of Kibana contributors we won't be able to address the issue via individual code reviews. An eslint rule might be a grind at the beginning to disable all lines with classes, but at least it would prevent new ones from being added. Otherwise a documentation update might be an easier alternative. From the text, it's not clear that the classes will be deleted soon and currently should not be used in production, maybe a deprecation could be added there?

yuliacech commented:
Hi @majagrubic, thanks a lot for your review! I addressed your feedback in https://github.com/elastic/kibana/pull/143194/commits/0d46ae3819d3303b98d4b38e366ea0dcd5d3b94e, could you please have another look?

kellyemurphy commented on packages/kbn-guided-onboarding/src/components/landing_page/observability_link_card.tsx:
```diff
@@ -0,0 +1,84 @@
+        'guidedOnboardingPackage.gettingStarted.linkCard.observability.cardDescription',
+        {
+          defaultMessage:
+            'Bring together application, infrastructure, and user data into a unified solution for end-to-end observability and alerting.',
```
```suggestion
            'Add application, infrastructure, and user data through our pre-built integrations.',
```

Reworked this text with the other UX writers to 1. introduce the idea of what an integration is, 2. shorten the text so it's closer to the other cards, 3. remove the repetition from the other cards.

kellyemurphy replied:
Otherwise, 👍 on the rest of the text.

kellyemurphy commented:
Looking back at the design files, there are more differences in the text here.
<img width="1256" alt="Screen Shot 2022-10-19 at 11 21 43 AM" src="https://user-images.githubusercontent.com/18013151/196749572-9d21607a-2999-4916-a4b9-e93d1f24cb00.png">

yuliacech commented:
Thanks a lot, @kellyemurphy! I used the figma file you linked and updated the texts, the screenshots in the PR description should be the latest state as well.

alisonelizabeth commented:
@elasticmachine merge upstream

majagrubic reviewed (approved):
Thanks for addressing the comments. Home plugin changes look good.

clintandrewhall commented on packages/kbn-guided-onboarding/src/components/landing_page/guide_card_footer.tsx:
```diff
@@ -0,0 +1,114 @@
+        // Used for FS tracking
+        data-test-subj={`onboarding--guideCard--view--${useCase}`}
+        fill
+        onClick={() => activateGuide(useCase, guideState)}
```
Incidentally, this is also a problem: you're creating a new function on every render cycle.

clintandrewhall commented on packages/kbn-guided-onboarding/src/components/landing_page/guide_card_footer.tsx:
```diff
@@ -0,0 +1,114 @@
+          // Used for FS tracking
+          data-test-subj={`onboarding--guideCard--continue--${useCase}`}
+          fill
+          onClick={() => activateGuide(useCase, guideState)}
```
Same.

clintandrewhall commented on packages/kbn-guided-onboarding/src/components/landing_page/guide_card_footer.tsx:
```diff
@@ -0,0 +1,114 @@
+  activateGuide: (useCase: UseCase, guideState?: GuideState) => void;
+}
+export const GuideCardFooter = ({ guides, useCase, activateGuide }: GuideCardFooterProps) => {
+  const guideState = guides.find((guide) => guide.guideId === (useCase as GuideId));
```
This feels like a code smell-- casting a `UseCase` as a `GuideId`, and that `UseCase` and `GuideId` are identical...?

clintandrewhall commented on packages/kbn-guided-onboarding/src/components/landing_page/use_case_card.tsx:
```diff
@@ -0,0 +1,105 @@
+  },
+};
+
+export type UseCase = 'search' | 'observability' | 'security';
```
I'm curious why this type is necessary, when it matches `GuideId`.  Appears to me these types need coalescing.

clintandrewhall reviewed (commented):
A few other comments while looking over this PR after the fact.  YMMV.  Feel free to reach out if Global Experience can help in any way!
